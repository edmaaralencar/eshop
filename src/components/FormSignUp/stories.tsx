import { StoryObj, Meta } from '@storybook/react'
import FormSignIn from '.'

export default {
  title: 'FormSignIn',
  component: FormSignIn
} as Meta

export const Default: StoryObj = {}
