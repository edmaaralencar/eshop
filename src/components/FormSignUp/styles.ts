import { darken } from 'polished'
import styled, { css, DefaultTheme } from 'styled-components'

export const Wrapper = styled.form`
  display: flex;
  flex-direction: column;

  width: 100%;
`

export const InputContainer = styled.div`
  display: flex;
  flex-direction: column;
  gap: 1.2rem;

  margin-bottom: 2.4rem;
`

export const FormGroup = styled.div`
  display: flex;
  flex-direction: column;
  gap: 0.8rem;

  span {
    color: #de3838;
    font-size: 1.2rem;
  }
`

type InputProps = {
  error: boolean
}

const inputModifier = {
  error: (theme: DefaultTheme) => css`
    border-color: ${theme.colors.red};

    &:focus-within {
      border-color: ${darken(0.1, theme.colors.red)};
    }
  `
}

export const Input = styled.div<InputProps>`
  ${({ theme, error }) => css`
    background-color: #ebebf1;
    display: flex;
    align-items: center;
    padding: 0 1rem;
    border: 0.2rem solid;
    border-color: #ebebf1;
    transition: border-color 0.2s ease-out;

    &:focus-within {
      border-color: ${theme.colors.primary.dark};
    }

    ${error && inputModifier.error(theme)}

    input {
      border: none;
      padding: 1.5rem 0;
      background: transparent;
      outline: none;
      padding-left: 1rem;
      font-size: 1.4rem;

      &::placeholder {
        font-size: 1.4rem;
      }
    }
  `}
`

export const AuthLink = styled.span`
  font-size: 1.4rem;
  display: block;
  text-align: center;
  margin-top: 3.2rem;

  a {
    color: ${({ theme }) => theme.colors.primary.dark};
    text-decoration: none;

    &:hover {
      color: ${({ theme }) => darken(0.1, theme.colors.primary.dark)};
    }
  }
`
