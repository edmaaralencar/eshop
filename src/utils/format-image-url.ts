export function formatImageUrl(image: string) {
  return `/products/${image}`
}
