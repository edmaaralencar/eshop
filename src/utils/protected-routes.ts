import { GetServerSidePropsContext } from 'next'
// import { unstable_getServerSession } from 'next-auth'
import { getSession } from 'next-auth/react'
// import { options } from 'pages/api/auth/[...nextauth].api'

async function protectedRoutes(context: GetServerSidePropsContext) {
  // const session = await unstable_getServerSession(
  //   context.req,
  //   context.res,
  //   options
  // )

  // if (!session) {
  //   return {
  //     redirect: {
  //       destination: `/sign-in?callbackUrl=${context.resolvedUrl}`,
  //       permanent: false
  //     }
  //   }
  // }

  const session = await getSession(context)

  if (!session) {
    context.res.setHeader(
      'Location',
      `/sign-in?callbackUrl=${context.resolvedUrl}`
    )
    context.res.statusCode = 302
  }

  return session
}

export default protectedRoutes
