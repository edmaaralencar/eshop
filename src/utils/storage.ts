const APP_KEY = '@eshop'

export function getStorageItem(key: string) {
  if (typeof window === 'undefined') return

  const data = window.localStorage.getItem(`${APP_KEY}.${key}`)
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  return JSON.parse(data!)
}

export function setStorageItem(key: string, value: any[]) {
  if (typeof window === 'undefined') return

  const data = JSON.stringify(value)
  return window.localStorage.setItem(`${APP_KEY}.${key}`, data)
}

export function removeStorageItem(key: string) {
  if (typeof window === 'undefined') return

  return window.localStorage.removeItem(`${APP_KEY}.${key}`)
}
