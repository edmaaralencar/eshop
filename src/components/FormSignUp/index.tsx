import Button from 'components/Button'
import Input from 'components/Input'
import useErrors from 'hooks/use-errors'
import { api } from 'lib/axios'
import Link from 'next/link'
import Router from 'next/router'
import { ChangeEvent, FormEvent, useState } from 'react'
import { FiLock, FiMail, FiUser } from 'react-icons/fi'
import { toast } from 'react-toastify'
import isEmailValid from 'utils/is-email-valid'
import * as S from './styles'

function FormSignUp() {
  const [name, setName] = useState('')
  const [email, setEmail] = useState('')
  const [password, setPassword] = useState('')

  const { setError, removeError, errors, getErrorMessageByFieldName } =
    useErrors()

  const isFormValid = name && email && password && errors.length === 0

  function handleNameChange(event: ChangeEvent<HTMLInputElement>) {
    setName(event.target.value)

    if (!event.target.value) {
      setError({ field: 'name', message: 'Nome é obrigatório.' })
    } else {
      removeError('name')
    }
  }

  function handleEmailChange(event: ChangeEvent<HTMLInputElement>) {
    setEmail(event.target.value)

    if (!event.target.value) {
      removeError('email-invalid')
      setError({ field: 'email-obligatory', message: 'E-mail é obrigatório.' })
      return
    }

    removeError('email-obligatory')

    if (!isEmailValid(event.target.value)) {
      setError({ field: 'email-invalid', message: 'E-mail é inválido.' })
    } else {
      removeError('email-invalid')
    }
  }

  function handlePasswordChange(event: ChangeEvent<HTMLInputElement>) {
    setPassword(event.target.value)

    if (!event.target.value) {
      setError({
        field: 'password',
        message: 'Senha é obrigatória.'
      })
    } else {
      removeError('password')
    }
  }

  async function handleSubmit(event: FormEvent) {
    event.preventDefault()

    const values = {
      name,
      email,
      password
    }

    try {
      const response = await api.post('/users/sign-up', values)
      console.log(response)
      toast.success('Conta criada com sucesso.')
      Router.push('/sign-in')
    } catch (error) {
      console.log(error)
    }
  }

  return (
    <S.Wrapper onSubmit={handleSubmit} noValidate>
      <S.InputContainer>
        <Input
          icon={<FiUser size={22} color="#2E2E2E" />}
          error={getErrorMessageByFieldName('name') as string}
          onChange={handleNameChange}
          value={name}
          placeholder="Nome*"
        />

        <Input
          icon={<FiMail size={22} color="#2E2E2E" />}
          error={
            [
              getErrorMessageByFieldName('email-invalid'),
              getErrorMessageByFieldName('email-obligatory')
            ] as string[]
          }
          onChange={handleEmailChange}
          value={email}
          placeholder="Email*"
        />

        <Input
          icon={<FiLock size={22} color="#2E2E2E" />}
          error={getErrorMessageByFieldName('password') as string}
          onChange={handlePasswordChange}
          value={password}
          placeholder="Senha*"
          type="password"
        />
      </S.InputContainer>
      <Button type="submit" disabled={!isFormValid} fullWidth>
        Entrar
      </Button>
      <S.AuthLink>
        Possui uma conta? <Link href="/sign-in">Faça login</Link>
      </S.AuthLink>
    </S.Wrapper>
  )
}

export default FormSignUp
