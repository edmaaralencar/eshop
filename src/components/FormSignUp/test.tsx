import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/test-helper'

import FormSignUp from '.'

describe('<FormSignUp />', () => {
  it('should render the heading', () => {
    const { container } = renderWithTheme(<FormSignUp />)

    expect(screen.getByPlaceholderText(/nome*/i)).toBeInTheDocument()
    expect(screen.getByPlaceholderText(/email*/i)).toBeInTheDocument()
    expect(screen.getByPlaceholderText(/senha*/i)).toBeInTheDocument()

    expect(screen.getByRole('button', { name: /entrar/i })).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })
})
