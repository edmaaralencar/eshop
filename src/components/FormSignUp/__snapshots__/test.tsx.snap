// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormSignUp /> should render the heading 1`] = `
.c4 {
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #6C6CFF;
  color: #FFFFFF;
  -webkit-transition: background-color 0.2s ease-out;
  transition: background-color 0.2s ease-out;
  padding: 0 2rem;
  height: 5.2rem;
  width: 100%;
}

.c4 span {
  font-weight: 300;
}

.c4:disabled {
  cursor: not-allowed;
}

.c4:hover {
  background-color: #7676ff;
}

.c4 span {
  font-size: 1.8rem;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.8rem;
}

.c2 span {
  color: #de3838;
  font-size: 1.2rem;
}

.c3 {
  background-color: #ebebf1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 1rem;
  border: 0.2rem solid;
  border-color: #ebebf1;
  -webkit-transition: border-color 0.2s ease-out;
  transition: border-color 0.2s ease-out;
}

.c3:focus-within {
  border-color: #6C6CFF;
}

.c3 input {
  border: none;
  padding: 1.5rem 0;
  background: transparent;
  outline: none;
  padding-left: 1rem;
  font-size: 1.4rem;
}

.c3 input::-webkit-input-placeholder {
  font-size: 1.4rem;
}

.c3 input::-moz-placeholder {
  font-size: 1.4rem;
}

.c3 input:-ms-input-placeholder {
  font-size: 1.4rem;
}

.c3 input::placeholder {
  font-size: 1.4rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 1.2rem;
  margin-bottom: 2.4rem;
}

.c5 {
  font-size: 1.4rem;
  display: block;
  text-align: center;
  margin-top: 3.2rem;
}

.c5 a {
  color: #6C6CFF;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c5 a:hover {
  color: #3939ff;
}

<form
  class="c0"
  novalidate=""
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <svg
          color="#2E2E2E"
          fill="none"
          height="22"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          style="color: rgb(46, 46, 46);"
          viewBox="0 0 24 24"
          width="22"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M20 21v-2a4 4 0 0 0-4-4H8a4 4 0 0 0-4 4v2"
          />
          <circle
            cx="12"
            cy="7"
            r="4"
          />
        </svg>
        <input
          placeholder="Nome*"
          value=""
        />
      </div>
    </div>
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <svg
          color="#2E2E2E"
          fill="none"
          height="22"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          style="color: rgb(46, 46, 46);"
          viewBox="0 0 24 24"
          width="22"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4 4h16c1.1 0 2 .9 2 2v12c0 1.1-.9 2-2 2H4c-1.1 0-2-.9-2-2V6c0-1.1.9-2 2-2z"
          />
          <polyline
            points="22,6 12,13 2,6"
          />
        </svg>
        <input
          placeholder="Email*"
          value=""
        />
      </div>
      <span />
    </div>
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <svg
          color="#2E2E2E"
          fill="none"
          height="22"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          style="color: rgb(46, 46, 46);"
          viewBox="0 0 24 24"
          width="22"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="11"
            rx="2"
            ry="2"
            width="18"
            x="3"
            y="11"
          />
          <path
            d="M7 11V7a5 5 0 0 1 10 0v4"
          />
        </svg>
        <input
          placeholder="Senha*"
          type="password"
          value=""
        />
      </div>
    </div>
  </div>
  <button
    class="c4"
    disabled=""
    type="submit"
  >
    <span>
      Entrar
    </span>
  </button>
  <span
    class="c5"
  >
    Possui uma conta? 
    <a
      href="/sign-in"
    >
      Fa√ßa login
    </a>
  </span>
</form>
`;
